File: python/461.汉明距离.py
#
# @lc app=leetcode.cn id=461 lang=python3
#
# [461] 汉明距离
#

# @lc code=start
class Solution:
    # 内置函数
    def hammingDistance(self, x: int, y: int) -> int:
        return bin(x ^ y).count('1')

    # Brian Kernighan 算法
    def hammingDistance1(self, x: int, y: int) -> int:
        distance = x ^ y
        count = 0
        while distance:
            count += 1
            distance &= distance - 1
        return count
# @lc code=end

File: cpp/461.汉明距离.cpp
/*
 * @lc app=leetcode.cn id=461 lang=cpp
 *
 * [461] 汉明距离
 */
#include <algorithm>
#include <iostream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;
// @lc code=start
class Solution {
public:
    /**
     * Brian Kernighan 算法
    */
    int hammingDistance(int x, int y) {
        int count = 0;
        int n = x ^ y;
        while (n) {
            n &= n - 1;
            count++;
        }
        return count;
    }

    /**
     * O(log(n))
    */
    int hammingDistance2(int x, int y) {
        int count = 0;
        int n = x ^ y;
        while (n) {
            count += n & 1;
            n >>= 1;
        }
        return count;
    }

    /**
     * 自写,O(n)
    */
   int hammingDistance1(int x, int y) {
        int count = 0;
        for (int i = 0; i < 32; i++) {
            count = ((1 & x) ^ (1 & y)) ? count + 1 : count;
            x >>= 1;
            y >>= 1;
        }
        return count;
   }
};
// @lc code=end

